### PR TITLE
Simulink integration 

### PR DESCRIPTION
**Pull request by sthelia**
I've been working on improving an S-function to be able to integrate JSBSim with Simulink. 

I included an example Simulink file of running one of the 737 Scripts. It should work for both Windows and Linux users.

### CODE REVIEW

**Comment by bcoconni** [comment]
Thanks for the PR @sthelia :+1: 

As per [our discussion](https://github.com/JSBSim-Team/jsbsim/discussions/405#discussioncomment-801468) I have just pushed a commit (sthelia/JSBSim@d0cf1e2) to your PR to trigger the build of the S-Function by our CI/CD workflow (GitHub Actions).

This commit also allows the S-Function to be built along with the JSBSim C++ library: for that you need to pass `-DBUILD_MATLAB_SFUNCTION=ON` to CMake.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:572]:
```suggestion
				return false;
```
C++ uses `true` and `false` for booleans.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:578]:
```suggestion
					return false;
```
C++ uses `true` and `false` for booleans.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:586]:
```suggestion
					return false;
```
C++ uses `true` and `false` for booleans.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:591]:
```suggestion
					return false;
```
C++ uses `true` and `false` for booleans.

**Review by bcoconni — CHANGES_REQUESTED** [review]
Thanks for your contribution.

There are a number of minor modifications which needs to be addressed before merging this PR. Most of them are related to the fact that C++ uses `true` and `false` for booleans (not `0` and `1`).

**Comment by seanmcleod** [comment]
@sthelia next time use the 'Add suggestion to batch' instead of 'Commit Suggestion' for every little change. So that there is a single commit instead of a 100, and we don't get spammed with 100 emails 😉, one for each commit.

File: matlab/JSBSimInterface.cpp
Comment (sthelia) [reply matlab/JSBSimInterface.cpp:46]: What would you recommend doing here? 

A current issue is that i have to run clearSF.m after each Simulink simulation, and I think it is because the destructor isn't working properly. I am not sure how to resolve that issue however.

**Comment by sthelia** [comment]
> @sthelia next time use the 'Add suggestion to batch' instead of 'Commit Suggestion' for every little change. So that there is a single commit instead of a 100, and we don't get spammed with 100 emails 😉, one for each commit.

Oh sorry! Didn't know that was an option.

File: matlab/JSBSimInterface.cpp
Comment (seanmcleod) [reply matlab/JSBSimInterface.cpp:46]: What about calling `FDMExec.ResetToInitialConditions(mode)` in `mdlInitializeConditions()`?

**Comment by bcoconni** [comment]
>> @sthelia next time use the 'Add suggestion to batch' instead of 'Commit Suggestion' for every little change. So that there is a single commit instead of a 100, and we don't get spammed with 100 emails 😉, one for each commit.
>
> Oh sorry! Didn't know that was an option. 

I did not know about that either. Thanks for the tip @seanmcleod 👍

**Comment by bcoconni** [comment]
Yes C++ happened 😊

A couple of errors need to be fixed because the compiler is unhappy.

**Comment by bcoconni** [comment]
@sthelia I just pushed a commit to your PR to fix the C++ silliness that made the tests fail. I hope you don't mind that I have proceeded that way but I felt it would help resolving the errors faster than going through GitHub review process and waiting for your feedback. Especially because the problem was caused by C++ ~silly~ ~overzealous~ ~pedantic~ uncompromising syntax that I personally find quite uninteresting to discuss about.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [reply matlab/JSBSimInterface.cpp:46]: I think the problem lies in the variable `FDMExec` which you made a global variable in the file `JSBSim_SFunction.cpp` and as such "survives" as long as the S-Function exists; hence the need to clear everything. Given that `FDMExec` is only used by the class `JSBSimInterface` I would suggest a different approach and make the variable `FDMExec` a member of the `JSBSimInterface` class.

You also need to delete the instance of `JSBSimInterface` in `mdlTerminate`. At the moment, since the instance is not deleted your SFunction is leaking memory.

**Review by bcoconni — APPROVED** [review]

File: matlab/JSBSim_SFunction.cpp
Comment (bcoconni) [inline matlab/JSBSim_SFunction.cpp:159]:
```suggestion
```
Don't declare `FDMExec` a global variable. You'd rather make it a member of `JSBSimInterface`

File: matlab/JSBSim_SFunction.cpp
Comment (bcoconni) [inline matlab/JSBSim_SFunction.cpp:312]:
```suggestion
	    ssGetPWork(S)[0] = (void *) new JSBSimInterface(delta_t);
```
Make `FDMExec` a member of `JSBSimInterface`. It is not used anywhere else in the code `JSBSim_SFunction.cpp` anyway.

File: matlab/JSBSim_SFunction.cpp
Comment (bcoconni) [inline matlab/JSBSim_SFunction.cpp:593]:
```suggestion
	JSBSimInterface *JII = (JSBSimInterface *) ssGetPWork(S)[0];   // retrieve C++ object pointers vector
        delete JII; 
```
You must delete the instance of `JSBSimInterface` otherwise you'll be leaking memory.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:13]:
```suggestion
	fdmExec = new FGFDMExec;
```
Create an instance of `FGFDMExec` that will de destroyed as soon as `JSBSimInterface` is deleted.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:27]:
```suggestion
	fdmExec = new FGFDMExec;
```
Create an instance of `FGFDMExec` that will de destroyed as soon as `JSBSimInterface` is deleted.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:10]:
```suggestion
JSBSimInterface::JSBSimInterface(void)
```
The instance of `FGFDMExec` is now a member of `JSBSimInterface`, there is no need to pass it as an argument.

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [inline matlab/JSBSimInterface.cpp:24]:
```suggestion
JSBSimInterface::JSBSimInterface(double dt)
```
The instance of `FGFDMExec` is now a member of `JSBSimInterface`, there is no need to pass it as an argument.

File: matlab/JSBSimInterface.h
Comment (bcoconni) [inline matlab/JSBSimInterface.h:17]:
```suggestion
  JSBSimInterface(void);
  JSBSimInterface(double dt);
```
`FGFDMExec` instances will be now be a member of `JSBSimInterface`

File: matlab/JSBSimInterface.cpp
Comment (bcoconni) [reply matlab/JSBSimInterface.cpp:46]: Once you have made `FGFDMExec` a member of `JSBSimInterface`, you can call `delete` on `fdmExec`

**Review by bcoconni — COMMENTED** [review]
@sthelia I have initiated a new review regarding the problem you have reported about the S-Function not cleaning up after itself.

If you feel more comfortable with a commit review, I have compiled all the changes I am suggesting in bcoconni/jsbsim@0d1410b935aea7ed51e3ff3adc6264868f0f5480. Unfortunately, I don't have access to Matlab so you'll have to check yourself whether or not the changes I have suggested allow avoiding calling `clearSF.m` after each Simulink run. All I could check with gitHub Actions was that the code compiles successfully and that the example `ex737cruise.slx` runs usccessfully as well.

I might also push the changes directly to your PR: if you prefer so, just let me know.

**Comment by bcoconni** [comment]
@sthelia Also please include an open source license to all files in this PR otherwise we won't be able to pull them to JSBSim.

**Comment by agodemar** [comment]
@bcoconni Do someone of us admins have to "Squash and merge" this PR?

**Comment by seanmcleod** [comment]
@agodemar it looks like @bcoconni is waiting for feedback on his suggested changes for the S-function cleanup issue and an open source license for the code.

**Comment by sthelia** [comment]
I am currently on vacation and will not have access to my computer until next week - hope to test everything you have suggested then!

**Comment by bcoconni** [comment]
> I am currently on vacation and will not have access to my computer until next week - hope to test everything you have suggested then!

@sthelia No rush 😄 JSBSim has been lacking an S-Function for years, it can still wait for a few more days/weeks !
Enjoy your vacation !

**Comment by bcoconni** [comment]
@agodemar If you are eager to test the S-Function code, you can pull the complete code from my fork: bcoconni/jsbsim@0d1410b935aea7ed51e3ff3adc6264868f0f5480 (branch `pr/sthelia/445`). [GitHub Actions report that it compiles and executes the example `ex737cruise.slx` successfully](https://github.com/bcoconni/jsbsim/runs/2914006957?check_suite_focus=true).

**Comment by sthelia** [comment]
> @sthelia Also please include an open source license to all files in this PR otherwise we won't be able to pull them to JSBSim.

What should be included? Is it the copyright text mentioning GNU that is added on top on the other source files?

**Comment by sthelia** [comment]
Thank you @bcoconni for the help with the C++ sillyness fixes, I am more than happy to not have to understand why they happen :D I also appreciate the comments to improve and cleanup the code, I am definitely learning a lot from this.

**Comment by bcoconni** [comment]
> What should be included? Is it the copyright text mentioning GNU that is added on top on the other source files?

Well, you've entered the realm of licenses which is always a joy when dealing with a mixture of open source (JSBSim and your code) and proprietary (Matlab) licenses. From what I could gather, you stood on the shoulder of giants :wink: to write this code so we need to give them proper credits:

* Given that you used (and explicitly referenced) code from https://fr.mathworks.com/matlabcentral/fileexchange/25042-jsbsim-s-function-gui-0-3 and that Brian Mills has set an open source license to his code, you must abide to it:
```text
Copyright (c) 2009, Brian Mills
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
We also need to ask @podhrmic (Michal Podhradsky) for permission to use [his code](https://github.com/podhrmic/JSBSim-Matlab) under the license above since you explicitly mentioned Michal's code as a reference.

Finally, Brian Mills also mentioned @agodemar's work (Agostino De Marco) as the source for his original code but I guess this permission should be quite easy to get :wink:

@podhrmic and @agodemar, just drop a message below in this PR and I guess this would do as an evidence.

Once we've got due permission, you can copy the license term at the top of each file in this PR:
```c++
/*
Copyright (c) 2009, Brian Mills
All rights reserved.

Copyright (c) 2021, Agostino De Marco, Michal Podhradsky, Tilda Sikström

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
```

**Comment by bcoconni** [comment]
Regarding the files that need a license term, I'd say that you can limit yourself to the following non trivial files:
* `matlab/JSBSimInterface.cpp`
* `matlab/JSBSimInterface.h`
* `matlab/JSBSim_SFunction.cpp`
* `matlab/readme.txt`

The other files are either build files or trivial files so you can save yourself the trouble of putting a license in them.

**Comment by bcoconni** [comment]
Oh I just saw that @podhrmic gave credit to @EliaTarasov in [his repo](https://github.com/podhrmic/JSBSim-Matlab) then we also need Elia's permission.

**Comment by agodemar** [comment]
> Finally, Brian Mills also mentioned @agodemar's work (Agostino De Marco) as the source for his original code but I guess this permission should be quite easy to get 😉
> 
> @podhrmic and @agodemar, just drop a message below in this PR and I guess this would do as an evidence.

You have my permission. All this effort, up to this PR by @sthelia began from my original code back in 2009.

**Comment by EliaTarasov** [comment]
Feel free to use it, @bcoconni :)

**Comment by podhrmic** [comment]
Hi @bcoconni - you have my permission to use my code:-)

**Comment by sthelia** [comment]
Alright, the license term is added. 

Thanks again everyone who developed the code I based this on, I would not have been able to present this pull request without it!

**Comment by SmartUpEngineering** [comment]
> Alright, the license term is added.
> 
> Thanks again everyone who developed the code I based this on, I would not have been able to present this pull request without it!

@sthelia thank you for your contribution!

**Comment by bcoconni** [comment]
The PR is merged ! JSBSim now has its official interface with Matlab.
Thanks @sthelia for your commitment and bringing this contribution to its positive conclusion.
Thanks also to @agodemar, @EliaTarasov, @podhrmic for their work without which this PR would not exist today.

**Comment by agodemar** [comment]
> The PR is merged ! JSBSim now has its official interface with Matlab.
> Thanks @sthelia for your commitment and bringing this contribution to its positive conclusion.
> Thanks also to @agodemar, @EliaTarasov, @podhrmic for their work without which this PR would not exist today.

This is great news. Thank you @sthelia for your effort!

**Pull request merged**
